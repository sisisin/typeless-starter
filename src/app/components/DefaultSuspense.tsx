import React from 'react';

interface DefaultSuspenseProps {
  children: React.ReactNode;
}

export function DefaultSuspense(props: DefaultSuspenseProps) {
  const { children } = props;
  return <React.Suspense fallback="loading...">{children}</React.Suspense>;
}
