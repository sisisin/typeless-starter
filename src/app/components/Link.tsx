import { useRouter } from 'app/hooks/useRouter';
import { AppPaths, GetOptionFromPath } from 'app/types/AppRouteDefinitions';
import { LocationSource } from 'app/types/location';
import React from 'react';
// eslint-disable-next-line no-restricted-imports
import { Link as RouterLink } from 'react-router-dom';

export function TypedLink<T extends AppPaths>(
  props: React.PropsWithChildren<{ path: T } & GetOptionFromPath<T>>,
): JSX.Element;
export function TypedLink({ children, ...option }: React.PropsWithChildren<LocationSource>) {
  const { history } = useRouter();
  const to = history.createHref(option);
  return <RouterLink to={to}>{children}</RouterLink>;
}

export function AnyLink(props: React.PropsWithChildren<LocationSource>): JSX.Element;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function AnyLink({ children, ...option }: React.PropsWithChildren<any>) {
  return <TypedLink {...option}>{children}</TypedLink>;
}
