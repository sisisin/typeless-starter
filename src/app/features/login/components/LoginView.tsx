import React from 'react';
import { useActions } from 'typeless';
import { LoginActions } from '../interface';

export const LoginView = () => {
  const { loginSubmitted } = useActions(LoginActions);
  return (
    <div>
      <div>login!</div>
      <button onClick={loginSubmitted}>submit</button>
    </div>
  );
};
