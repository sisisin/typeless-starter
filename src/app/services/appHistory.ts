import { createBrowserHistory, History, LocationState } from 'history';

class AppHistory {
  constructor(private history: History<LocationState>) {}

  push(path: History.Path) {
    this.history.push(path);
  }

  toRouterProps() {
    return { history: this.history };
  }
}

export const appHistory = new AppHistory(createBrowserHistory());
