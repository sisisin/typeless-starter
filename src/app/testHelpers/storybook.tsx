import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { Registry, TypelessContext } from 'typeless';

export const AppProvider: React.FC = (props) => {
  return (
    <TypelessContext.Provider value={{ registry: new Registry() }}>
      <MemoryRouter>{props.children}</MemoryRouter>
    </TypelessContext.Provider>
  );
};
