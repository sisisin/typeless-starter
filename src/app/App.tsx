import React from 'react';
import { Router } from 'react-router-dom';
import { AppRoutes } from './components/AppRoutes';
import { useSessionModule } from './features/session/module';
import { appHistory } from './services/appHistory';

export const App: React.FC = () => {
  useSessionModule();

  return (
    <Router {...appHistory.toRouterProps()}>
      <AppRoutes></AppRoutes>
    </Router>
  );
};
