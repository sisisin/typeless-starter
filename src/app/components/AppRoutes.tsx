import loadable, { LoadableComponent } from '@loadable/component';
import { useIsLoggedIn } from 'app/features/session/selector';
import React from 'react';
import { Redirect, Route, RouteProps, Switch } from 'react-router-dom';
import { HeaderMenu } from './HeaderMenu';

const SampleModule = loadable(() => import('app/features/sample/module'));
const LoginModule = loadable(() => import('app/features/login/module'));

type Base = {
  path: string | string[];
  requiresAuth: boolean;
};
type AppRouteProps = Base & {
  Component: React.ComponentType;
};
type AppLazyRouteProps = Base & {
  Component: LoadableComponent<unknown>;
};

export const lazyRoutes: AppLazyRouteProps[] = [
  { path: '/sample', requiresAuth: true, Component: SampleModule },
  { path: '/sample/:id', requiresAuth: true, Component: SampleModule },
  { path: '/login', requiresAuth: false, Component: LoginModule },
];
const routes: AppRouteProps[] = [
  {
    path: '/authed',
    requiresAuth: true,
    Component: () => <div>ok!</div>,
  },
];

const useRouteDefinitions = () => {
  const isLoggedIn = useIsLoggedIn();

  return routes.concat(lazyRoutes).map(({ requiresAuth, path, Component }, key) => {
    const base: RouteProps & { key: React.Key } = { key, exact: true, sensitive: true, path };

    if (requiresAuth) {
      return (
        <Route
          {...base}
          render={() => (isLoggedIn ? <Component></Component> : <Redirect key={key} to="/login"></Redirect>)}
        ></Route>
      );
    } else {
      return <Route {...base} component={Component}></Route>;
    }
  });
};

export const AppRoutes: React.FC = () => {
  return (
    <>
      <HeaderMenu></HeaderMenu>
      <Switch>
        {useRouteDefinitions()}
        <Route path="*">not found...</Route>
      </Switch>
    </>
  );
};
