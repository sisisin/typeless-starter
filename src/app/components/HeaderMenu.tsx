import React from 'react';
import { TypedLink, AnyLink } from 'app/components/Link';

export const HeaderMenu = () => {
  return (
    <>
      <div>
        <TypedLink path="/sample">sample </TypedLink>
      </div>
      <div>
        <TypedLink path="/login">login page</TypedLink>
      </div>
      <div>
        <AnyLink path="/authed">authed</AnyLink>
      </div>
      <div>
        <AnyLink path="/no_matched">no_matched</AnyLink>
      </div>
      <hr></hr>
    </>
  );
};
