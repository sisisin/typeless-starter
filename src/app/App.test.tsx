import { lazyRoutes } from 'app/components/AppRoutes';
import { TestProvider } from 'app/testHelpers/storybook';
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { App } from './App';
import { appHistory } from './services/appHistory';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <TestProvider>
      <App />
    </TestProvider>,
    div,
  );
  ReactDOM.unmountComponentAtNode(div);
  expect(true).toBeTruthy();
});

it('renders login', async () => {
  const node = renderer.create(
    <TestProvider>
      <App></App>
    </TestProvider>,
  );
  appHistory.push('/login');
  await lazyRoutes.find(({ path }) => path === '/login')!.Component.load();
  expect(node!.toJSON()).toMatchSnapshot();
});
