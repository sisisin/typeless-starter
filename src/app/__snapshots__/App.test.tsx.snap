// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders login 1`] = `
Array [
  <div>
    <a
      href="/sample"
      onClick={[Function]}
    >
      sample 
    </a>
  </div>,
  <div>
    <a
      href="/login"
      onClick={[Function]}
    >
      login page
    </a>
  </div>,
  <div>
    <a
      href="/authed"
      onClick={[Function]}
    >
      authed
    </a>
  </div>,
  <div>
    <a
      href="/no_matched"
      onClick={[Function]}
    >
      no_matched
    </a>
  </div>,
  <hr />,
  <div>
    <div>
      login!
    </div>
    <button
      onClick={[Function]}
    >
      submit
    </button>
  </div>,
]
`;
